pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -926,49 +966,93 @@ func (service *SecurityPolicyService) markDeleteShares(existingShares []*model.S
 		if err != nil {
 			log.Error(err, "Failed to wrap NSX project groups and shares", "nsxSecurityPolicyId", nsxSecurityPolicy.Id)
-			return nsxGetSecurityPolicy, err
+			return nil, err
 		}
 	}
 
 	// Wrap SecurityPolicy, groups, rules under VPC level together with project groups and shares into one hierarchy resource tree.
 	orgRoot, err := service.wrapHierarchyVpcSecurityPolicy(nsxSecurityPolicy, nsxGroups, projectInfraResource, vpcInfo)
 	if err != nil {
-		log.Error(err, "Failed to wrap SecurityPolicy in VPC", "nsxSecurityPolicyId", nsxSecurityPolicy.Id)
-		return nsxGetSecurityPolicy, err
+		log.Error(err, "Failed to wrap NSX SecurityPolicy resources in VPC", "nsxSecurityPolicyId", nsxSecurityPolicy.Id)
+		return nil, err
 	}
-
-	// Create/update or delete SecurityPolicy together with groups, rules under VPC level and project groups, shares.
+	// Create/update SecurityPolicy together with groups, rules under VPC level and project groups, shares.
 	err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam)
 	err = nsxutil.TransNSXApiError(err)
 	if err != nil {
-		log.Error(err, "Failed to create/update or delete SecurityPolicy in VPC", "nsxSecurityPolicyId", nsxSecurityPolicy.Id)
-		return nsxGetSecurityPolicy, err
+		log.Error(err, "Failed to create or update NSX SecurityPolicy in VPC", "nsxSecurityPolicyId", nsxSecurityPolicy.Id)
+		return nil, err
 	}
 
-	if !isDelete {
-		// Get SecurityPolicy from NSX after HAPI call as NSX renders several fields like `path`/`parent_path`.
-		nsxGetSecurityPolicy, err = service.NSXClient.VPCSecurityClient.Get(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, *nsxSecurityPolicy.Id)
-		err = nsxutil.TransNSXApiError(err)
+	// Get SecurityPolicy from NSX after HAPI call as NSX renders several fields like `path`/`parent_path`.
+	nsxGetSecurityPolicy, err := service.NSXClient.VPCSecurityClient.Get(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, *nsxSecurityPolicy.Id)
+	err = nsxutil.TransNSXApiError(err)
+	if err != nil {
+		log.Error(err, "Failed to get NSX SecurityPolicy in VPC", "nsxSecurityPolicyId", nsxSecurityPolicy.Id)
+		return nil, err
+	}
+
+	return &nsxGetSecurityPolicy, nil
+}
+
+// deleteNSXSecurityPolicy deletes NSX SecurityPolicy, rules and the groups/shares for non-Default Project.
+func (service *SecurityPolicyService) deleteNSXSecurityPolicy(nsxSecurityPolicy *model.SecurityPolicy, vpcInfo *common.VPCResourceInfo) error {
+	var err error
+
+	// Delete NSX SecurityPolicy and rules only.
+	err = service.NSXClient.VPCSecurityClient.Delete(vpcInfo.OrgID, vpcInfo.ProjectID, vpcInfo.VPCID, *nsxSecurityPolicy.Id)
+	err = nsxutil.TransNSXApiError(err)
+	if err != nil {
+		log.Error(err, "Failed to delete NSX SecurityPolicy in VPC", "nsxSecurityPolicyId", nsxSecurityPolicy.Id)
+		return err
+	}
+
+	return nil
+}
+
+// deleteNSXSecurityPolicyGroupShare deletes NSX SecurityPolicy associated the groups/shares for non-Default Project.
```

Comment:
So you associated the deletions of groups, shares and shared groups... For example, if we want to delete a group A, a share B and a shared group C, but A is used in some user-created resource, then can B and C be deleted?